tests/kafkatest/services/monitor/http.py

```diff
@@ -0,0 +1,225 @@
+# for highly detailed messages, e.g. logging every single incoming request.
+TRACE = 5
+
+
+class HttpMetricsCollector(object):
+    """
+    HttpMetricsCollector enables collection of metrics from various Kafka clients instrumented with the
+    PushHttpMetricsReporter. It starts a web server locally and provides the necessary configuration for clients
+    to automatically report metrics data to this server. It also provides basic functionality for querying the
+    recorded metrics. This class can be used either as a mixin or standalone object.
+    """
+
+    # The port to listen on on the worker node, which will be forwarded to the port listening on this driver node
+    REMOTE_PORT = 6789
+
+    def __init__(self, **kwargs):
+        """
+        Create a new HttpMetricsCollector
+        :param period the period, in seconds, between updates that the metrics reporter configuration should define.
+               defaults to reporting once per second
+        :param args:
+        :param kwargs:
+        """
+        self._http_metrics_period = kwargs.pop('period', 1)
+
+        super(HttpMetricsCollector, self).__init__(**kwargs)
+
+        # TODO: currently we maintain just a simple map from all key info -> value. However, some key fields are far
+        # more common to filter on, so we'd want to index by them, e.g. host, client.id, metric name.
+        self._http_metrics = defaultdict(list)
+
+        self._httpd = HTTPServer(('', 0), _MetricsReceiver)
+        self._httpd.parent = self
+        self._httpd.metrics = self._http_metrics
+
+        self._http_metrics_thread = Thread(target=self._run_http_metrics_httpd,
+                                           name='http-metrics-thread[%s]' % str(self))
+        self._http_metrics_thread.start()
+
+        self._forwarders = {}
+
+    @property
+    def http_metrics_url(self):
+        """
+        :return: the URL to use when reporting metrics
+        """
+        return "http://%s:%d" % ("localhost", self.REMOTE_PORT)
+
+    @property
+    def http_metrics_client_configs(self):
+        """
+        Get client configurations that can be used to report data to this collector. Note that in some cases
+        (e.g. streams, connect) these settings may need to be prefixed
+        :return: a dictionary of client configurations that will direct a client to report metrics to this collector
+        """
+        return {
+            "metric.reporters": "org.apache.kafka.tools.PushHttpMetricsReporter",
+            "metrics.url": self.http_metrics_url,
+            "metrics.period": self._http_metrics_period,
+        }
+
+    def start_node(self, node):
+        local_port = self._httpd.socket.getsockname()[1]
+        self.logger.debug('HttpMetricsCollector listening on %s', local_port)
+        self._forwarders[self.idx(node)] = _ReverseForwarder(self.logger, node, self.REMOTE_PORT, local_port)
+
+        super(HttpMetricsCollector, self).start_node(node)
+
+    def stop(self):
```

Comment:
The `ProducerPerformanceService` would have to call `super.stop()` for this to be invoked right? I don't think it does it presently. Please correct me if I am wrong.

Reply:
It would need to if it overrode `stop()`. However, the method resolution order for multiple inheritance will make this run first (because the mixin is listed in the super classes of `ProducerPerformanceService` first), then this `super` call will pass it along. Here's a quick example showing the equivalent class hierarchy:

```
>>> class Base(object):
...     pass
...
>>> class Mixin(object):
...     pass
...
>>> class Concrete(Mixin, Base):
...     pass
...
>>> Concrete.mro()
[<class '__main__.Concrete'>, <class '__main__.Mixin'>, <class '__main__.Base'>, <type 'object'>]
```

When python dispatches the initial call and any `super` calls, it works in the order of that list.

Putting mixins first usually makes life easier since the mixins have to take care to properly call `super` but classes using the mixin often don't, or only need to for methods they would have to call it for under normal circumstances in order to let the super class do its work.

This is also why the constructor is written the way it is with the flexible `**kwargs` -- it allows the mixin to work with pretty much any other constructor parameters from the class using the mixin.

Reply:
Btw, the call order being confusing is a downside of mixins in python, I was also debating whether we should just use composition and have the classes using this do everything explicitly. The nice thing with this approach is that as long as this class implements start + clean up properly, there isn't any burden on the class using it.

Reply:
Thanks for that note ewen. I learned something!